### PR TITLE
Add orient variables on slice

### PR DESCRIPTION
**Pull request by osheamonn**

### CODE REVIEW

File: src/DataStructures/VariablesHelpers.hpp [hunk 1]
```diff
@@ -43,3 +47,157 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
...
+  }
+  return oriented_offsets;
+}
+
+// The slice of a 3D mesh can have eight different data-storage orders
+// depending upon the relative orientation of the neighbor.  These are
+// determined by whether the new data-storage order varies fastest by the
+// lowest dim or the highest dim, and by whether each axis is aligned or
+// anti-aligned.
+inline std::vector<size_t> oriented_offset(
```
Comment (kidder) [inline src/DataStructures/VariablesHelpers.hpp]: move to cpp

File: src/DataStructures/VariablesHelpers.hpp [hunk 2]
```diff
@@ -43,3 +47,157 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
...
+    const Index<2>& slice_mesh, const size_t sliced_dim,
+    const Orientation<3>& orientation_of_neighbor) {
+  const std::array<size_t, 2> dims_of_slice =
+      (0 == sliced_dim ? make_array(1_st, 2_st)
+                       : (1 == sliced_dim) ? make_array(0_st, 2_st)
+                                           : make_array(0_st, 1_st));
+  const bool transpose_needed =
+      (orientation_of_neighbor.mapped(dims_of_slice[0]) >
+       orientation_of_neighbor.mapped(dims_of_slice[1]));
+  const Direction<3> first_nbr_axis = orientation_of_neighbor.mapped(
```
Comment (kidder) [inline src/DataStructures/VariablesHelpers.hpp]: nbr -> neighbor

File: src/DataStructures/VariablesHelpers.hpp [hunk 3]
```diff
@@ -43,3 +47,157 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
...
+      Direction<3>(dims_of_slice[0], Side::Upper));
+  const Direction<3> second_nbr_axis = orientation_of_neighbor.mapped(
+      Direction<3>(dims_of_slice[1], Side::Upper));
+  const bool first_nbr_axis_flipped = (Side::Lower == first_nbr_axis.side());
```
Comment (kidder) [inline src/DataStructures/VariablesHelpers.hpp]: neighbor_first_axis

File: src/DataStructures/VariablesHelpers.hpp [hunk 4]
```diff
@@ -43,3 +47,157 @@ Variables<TagsList> data_on_slice(const Variables<TagsList>& vars,
...
+  const std::array<size_t, 2> dims_of_slice =
+      (0 == sliced_dim ? make_array(1_st, 2_st)
+                       : (1 == sliced_dim) ? make_array(0_st, 2_st)
+                                           : make_array(0_st, 1_st));
+  const bool transpose_needed =
+      (orientation_of_neighbor.mapped(dims_of_slice[0]) >
+       orientation_of_neighbor.mapped(dims_of_slice[1]));
+  const Direction<3> first_nbr_axis = orientation_of_neighbor.mapped(
+      Direction<3>(dims_of_slice[0], Side::Upper));
+  const Direction<3> second_nbr_axis = orientation_of_neighbor.mapped(
```
Comment (kidder) [inline src/DataStructures/VariablesHelpers.hpp]: neighbor_second_axis

File: tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+using AffineCoordMap2D =
+    CoordinateMap<Frame::Logical, Frame::Grid, AffineMap2D>;
+using AffineCoordMap3D =
+    CoordinateMap<Frame::Logical, Frame::Grid, AffineMap3D>;
+template <size_t Dim>
+using Rotation = CoordinateMaps::Rotation<Dim>;
+
+namespace Tags {
+template <size_t SpatialDim>
+struct PhysicalCoords : db::DataBoxTag {
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp]: don't need tag

File: tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+  using type = tnsr::I<DataVector, SpatialDim, Frame::Grid>;
+  static constexpr db::DataBoxString_t label = "PhysicalCoords";
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp]: don't need label

File: tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+};
+}  // namespace Tags
+
+namespace {
+template <size_t SpatialDim>
+void check_orient_variables_on_slice(
+    const Block<SpatialDim, Frame::Grid>& my_block,
+    const Block<SpatialDim, Frame::Grid>& nbr_block,
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp]: nbr -> neighbor

Comment (kidder) [reply tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp]: in the whole file!

File: tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+    const Direction<SpatialDim>& direction_to_nbr,
+    const Index<SpatialDim>& my_extents, const Index<SpatialDim>& nbr_extents) {
+  using Tag = Tags::PhysicalCoords<SpatialDim>;
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp]: Tag -> tag

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/DataStructures/VariablesHelpers.hpp [hunk 9]
```diff
@@ -57,113 +57,13 @@ inline std::vector<size_t> oriented_offset(
...
-          }
-        }
-      } else {
-        std::iota(oriented_offsets.begin(), oriented_offsets.end(), 0);
-      }
-    }
-  }
-  return oriented_offsets;
-}
+std::vector<size_t> oriented_offset(
```
Comment (nilsdeppe) [inline src/DataStructures/VariablesHelpers.hpp]: noexcept all of these

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
rebase and squash

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**